db/view/view_update_backlog.hh

```diff
@@ -27,6 +27,12 @@ struct update_backlog {
     size_t current;
     size_t max;
 
+    update_backlog(size_t current, size_t max)
+        : current(current), max(max) {
+            assert(max != 0);
```

Comment:
I don't think this is a good idea to just introduce this assert. This structure doesn't uphold the invariant that `max != 0`, the `max` field is public and anybody can set it to 0 afterwards. Now imagine that such an instance of the structure is serialized into gossiper state and then all nodes read it and initialize the struct via this constructor, causing a cluster-wide assertion failure.

You should either:

1. Go all the way and make `current`/`max` fields immutable,
2. Drop the assert and let it be. Although `relative_size()` divides by `max`, it's a floating point division, so it will return a NaN and it's not the end of the world,
3. Make it possible to modify the fields only via setters that uphold the invariants.

Option 1 would be the best. Option 2 also sounds OK, though in one place we convert the result of `relative_size()` to an integer which could lead to UB if it returns a NaN.

Another note: we should rather avoid `assert` here and use `on_internal_error` instead. See the discussion in https://github.com/scylladb/scylladb/issues/7871

Reply:
I tried implementing option 1 in the rebase, please take a look. For now I left the `assert` in - using `on_internal_error` requires providing a logger and currently the view update backlog implementation is kept entirely in the header, so we have a few options of fixing this:
1. get the logger from somewhere, e.g. `db/view.cc`, first making it non-static
2. move the implementation to a separate file and add a new logger there
3. move the implementation to an existing file and use its logger

Any preference out of these three ?(maybe we can also just leave the assert)

Reply:
> Any preference out of these three ?

I think the first option is OK.

> (maybe we can also just leave the assert)

No, `assert` behaves differently. We don't remove asserts in release mode, so if the condition is not satisfied then the node will crash. `on_internal_error` only will throw an exception (unless you set a flag in the config that causes a crash, but it only makes to do so in tests and not production).

Reply:
Ok, I changed the `assert` to an `on_internal_error` added an extern logger from `db/view/view.cc`